File: app/src/main/java/by/overpass/gather/data/repo/message/MessageRepo.java
package by.overpass.gather.data.repo.message;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.annimon.stream.Stream;
import by.overpass.gather.commons.android.lifecycle.LiveDataUtils;
import by.overpass.gather.data.repo.meeting.MeetingsMetadata;
import by.overpass.gather.data.repo.user.UserData;
import by.overpass.gather.ui.meeting.chat.DeleteStatus;
import by.overpass.gather.ui.meeting.chat.MessageStatus;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.Date;
import java.util.List;

import javax.annotation.Nullable;
import javax.inject.Inject;

public class MessageRepo implements MeetingsMetadata {

    private final FirebaseFirestore firestore;

    @Inject
    public MessageRepo(FirebaseFirestore firestore) {
        this.firestore = firestore;
    }

    public LiveData<MessageStatus> messages(String meetingId) {
        MutableLiveData<MessageStatus> messagesData = new MutableLiveData<>();
        firestore.collection(COLLECTION_MEETINGS)
                .document(meetingId)
                .collection(Messages.COLLECTION)
                .orderBy(Messages.FIELD_DATE, Query.Direction.DESCENDING)
                .addSnapshotListener(new MessageMapper() {
                    @Override
                    void onSuccess(MessageStatus.Success success) {
                        messagesData.setValue(success);
                    }

                    @Override
                    void onError(MessageStatus.Error error) {
                        messagesData.setValue(error);
                    }
                });
        return messagesData;
    }

    public void send(String meetingId, @NonNull String input, @NonNull UserData userData) {
        firestore.collection(COLLECTION_MEETINGS)
                .document(meetingId)
                .collection(Messages.COLLECTION)
                .add(new BaseMessage(
                        input,
                        new Date(),
                        userData.getId(),
                        userData.getName(),
                        userData.getPhotoUrl()
                ));
    }

    public LiveData<DeleteStatus> delete(String meetingId, List<String> ids) {
        List<LiveData<DeleteStatus>> separateDeletions = Stream.of(ids)
                .map(id -> delete(meetingId, id))
                .toList();
        LiveData<List<DeleteStatus>> deletionsData = LiveDataUtils.zip(separateDeletions);
        return Transformations.map(deletionsData, input -> {
            DeleteStatus deleteStatus = new DeleteStatus.Progress();
            int successCount = 0;
            for (DeleteStatus status : input) {
                if (status instanceof DeleteStatus.Error
                        || status instanceof DeleteStatus.Progress) {
                    deleteStatus = status;
                    break;
                } else {
                    successCount++;
                }
            }
            if (successCount == input.size()) {
                deleteStatus = new DeleteStatus.Success();
            }
            return deleteStatus;
        });
    }

    private LiveData<DeleteStatus> delete(String meetingId, String messageId) {
        MutableLiveData<DeleteStatus> deleteData = new MutableLiveData<>();
        deleteData.setValue(new DeleteStatus.Progress());
        firestore.collection(COLLECTION_MEETINGS)
                .document(meetingId)
                .collection(Messages.COLLECTION)
                .document(messageId)
                .delete()
                .addOnSuccessListener(__ -> {
                    deleteData.setValue(new DeleteStatus.Success());
                })
                .addOnFailureListener(e -> deleteData.setValue(new DeleteStatus.Error(e)));
        return deleteData;
    }

    private static abstract class MessageMapper implements EventListener<QuerySnapshot> {

        @Override
        public void onEvent(@Nullable QuerySnapshot snapshots,
                            @Nullable FirebaseFirestoreException e) {
            if (e != null) {
                onError(new MessageStatus.Error(e));
            }
            if (snapshots != null) {
                List<Message> messages = Stream.of(snapshots.getDocuments())
                        .map(this::toMessage)
                        .toList();
                onSuccess(new MessageStatus.Success(messages));
            }
        }

        abstract void onSuccess(MessageStatus.Success success);

        abstract void onError(MessageStatus.Error error);

        private Message toMessage(DocumentSnapshot documentSnapshot) {
            return new Message(
                    documentSnapshot.getId(),
                    documentSnapshot.getString(Messages.FIELD_TEXT),
                    documentSnapshot.getDate(Messages.FIELD_DATE),
                    documentSnapshot.getString(Messages.FIELD_AUTHOR_ID),
                    documentSnapshot.getString(Messages.FIELD_AUTHOR_NAME),
                    documentSnapshot.getString(Messages.FIELD_AUTHOR_PHOTO_URL)
            );
        }
    }
}

File: app/src/main/java/by/overpass/gather/data/repo/message/MessageRepo.kt
package by.overpass.gather.data.repo.message

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import by.overpass.gather.commons.android.lifecycle.LiveDataUtils
import by.overpass.gather.data.repo.meeting.MeetingsMetadata
import by.overpass.gather.data.repo.user.UserData
import by.overpass.gather.ui.meeting.chat.DeleteStatus
import by.overpass.gather.ui.meeting.chat.MessageStatus
import com.google.firebase.firestore.*
import com.google.firebase.firestore.EventListener
import java.util.*
import javax.inject.Inject

class MessageRepo @Inject constructor(
        private val firestore: FirebaseFirestore
) : MeetingsMetadata {

    fun messages(meetingId: String): LiveData<MessageStatus> {
        val messagesData = MutableLiveData<MessageStatus>()
        firestore.collection(MeetingsMetadata.COLLECTION_MEETINGS)
                .document(meetingId)
                .collection(MeetingsMetadata.Messages.COLLECTION)
                .orderBy(MeetingsMetadata.Messages.FIELD_DATE, Query.Direction.DESCENDING)
                .addSnapshotListener(object : MessageMapper() {
                    override fun onSuccess(success: MessageStatus.Success) {
                        messagesData.value = success
                    }

                    override fun onError(error: MessageStatus.Error) {
                        messagesData.value = error
                    }
                })
        return messagesData
    }

    fun send(meetingId: String, input: String, userData: UserData) {
        firestore.collection(MeetingsMetadata.COLLECTION_MEETINGS)
                .document(meetingId)
                .collection(MeetingsMetadata.Messages.COLLECTION)
                .add(BaseMessage(
                        input,
                        Date(),
                        userData.id,
                        userData.name,
                        userData.photoUrl
                ))
    }

    fun delete(meetingId: String, ids: List<String>): LiveData<DeleteStatus> {
        val separateDeletions = ids
                .map { id ->
                    delete(meetingId, id)
                }
        val deletionsData = LiveDataUtils.zip(separateDeletions)
        return Transformations.map(deletionsData) { input ->
            var deleteStatus: DeleteStatus = DeleteStatus.Progress()
            var successCount = 0
            for (status in input) {
                if (status is DeleteStatus.Error || status is DeleteStatus.Progress) {
                    deleteStatus = status
                    break
                } else {
                    successCount++
                }
            }
            if (successCount == input.size) {
                deleteStatus = DeleteStatus.Success()
            }
            deleteStatus
        }
    }

    private fun delete(meetingId: String, messageId: String): LiveData<DeleteStatus> {
        val deleteData = MutableLiveData<DeleteStatus>()
        deleteData.setValue(DeleteStatus.Progress())
        firestore.collection(MeetingsMetadata.COLLECTION_MEETINGS)
                .document(meetingId)
                .collection(MeetingsMetadata.Messages.COLLECTION)
                .document(messageId)
                .delete()
                .addOnSuccessListener { s -> deleteData.setValue(DeleteStatus.Success()) }
                .addOnFailureListener { e -> deleteData.setValue(DeleteStatus.Error(e)) }
        return deleteData
    }

    private abstract class MessageMapper : EventListener<QuerySnapshot> {

        override fun onEvent(snapshots: QuerySnapshot?,
                             e: FirebaseFirestoreException?) {
            if (e != null) {
                onError(MessageStatus.Error(e))
            }
            if (snapshots != null) {
                val messages = snapshots.documents
                        .map { toMessage(it) }
                onSuccess(MessageStatus.Success(messages))
            }
        }

        internal abstract fun onSuccess(success: MessageStatus.Success)

        internal abstract fun onError(error: MessageStatus.Error)

        private fun toMessage(documentSnapshot: DocumentSnapshot): Message {
            return Message(
                    documentSnapshot.id,
                    documentSnapshot.getString(MeetingsMetadata.Messages.FIELD_TEXT),
                    documentSnapshot.getDate(MeetingsMetadata.Messages.FIELD_DATE),
                    documentSnapshot.getString(MeetingsMetadata.Messages.FIELD_AUTHOR_ID),
                    documentSnapshot.getString(MeetingsMetadata.Messages.FIELD_AUTHOR_NAME),
                    documentSnapshot.getString(MeetingsMetadata.Messages.FIELD_AUTHOR_PHOTO_URL)
            )
        }
    }
}
